src/julia.h

```diff
@@ -149,17 +155,30 @@ typedef struct {
 
     // followed by alignment padding and inline data, or owner pointer
 } jl_array_t;
+
 // compute # of extra words needed to store dimensions
 STATIC_INLINE int jl_array_ndimwords(uint32_t ndims)
 {
     return (ndims < 3 ? 0 : ndims-2);
 }
 
+typedef struct _jl_datatype_t jl_tupletype_t;
+
+// TupleMap is an implicitly defined type
+// that can consist of any of the following nodes:
+//   typedef TupleMap Union{TupleMapLevel, TupleMapEntry, Void}
+// it consists of roughly tree-shaped nodes, when possible because `isleaftype(key)`
+// otherwise the leaf entries are stored sorted, linearly
```

Comment:
I find this comment pretty unclear. Need a more thorough description of the data structure.